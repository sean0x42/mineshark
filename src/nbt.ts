export enum NbtType {
  End = 0,
  Byte,
  Short,
  Int,
  Long,
  Float,
  Double,
  ByteArray,
  String,
  List,
  Compound,
  // IntArray,
  // LongArray,
}

interface BaseNbtTag {
  kind: NbtType;
  name?: string;
  payload: unknown;
}

export interface NbtTagByte extends BaseNbtTag {
  kind: NbtType.Byte;
  payload: number;
}

export interface NbtTagShort extends BaseNbtTag {
  kind: NbtType.Short;
  payload: number;
}

export interface NbtTagInt extends BaseNbtTag {
  kind: NbtType.Int;
  payload: number;
}

export interface NbtTagLong extends BaseNbtTag {
  kind: NbtType.Long;
  payload: bigint;
}

export interface NbtTagFloat extends BaseNbtTag {
  kind: NbtType.Float;
  payload: number;
}

export interface NbtTagDouble extends BaseNbtTag {
  kind: NbtType.Double;
  payload: number;
}

export interface NbtTagByteArray extends BaseNbtTag {
  kind: NbtType.ByteArray;
  payload: Buffer;
}

export interface NbtTagString extends BaseNbtTag {
  kind: NbtType.String;
  payload: string;
}

export interface NbtTagList extends BaseNbtTag {
  kind: NbtType.List;
  payload: Array<NbtTag>;
}

export interface NbtTagCompound extends BaseNbtTag {
  kind: NbtType.Compound;
  payload: Record<string, NbtTag>;
}

// export interface NbtTagIntArray extends BaseNbtTag {
//   kind: NbtType.IntArray;
//   payload: Array<number>;
// }

// export interface NbtTagLongArray extends BaseNbtTag {
//   kind: NbtType.LongArray;
//   payload: Array<BigInt>;
// }

export type NbtTag =
  | NbtTagByte
  | NbtTagShort
  | NbtTagInt
  | NbtTagLong
  | NbtTagFloat
  | NbtTagDouble
  | NbtTagByteArray
  | NbtTagString
  | NbtTagList
  | NbtTagCompound;
// | NbtTagIntArray
// | NbtTagLongArray;
