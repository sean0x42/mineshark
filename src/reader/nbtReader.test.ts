import { NbtType } from "../nbt";
import NbtPacketReader from "./nbtReader";
import helloWorldNbt from "./__fixtures__/helloWorld.json";
import bigTestNbt from "./__fixtures__/bigTest.json";

describe("nbtReader", () => {
  // Note: creates a buffer in the uncompressed format.
  function createDummyBuffer(vals: number[]): Buffer {
    return Buffer.from([0x00, 0x00, ...vals]);
  }

  it("should parse the hello_world.nbt test case", () => {
    // Given
    const packet = createDummyBuffer(helloWorldNbt);

    // When
    const reader = new NbtPacketReader(packet);
    const nbt = reader.readNbtCompound();

    // Then
    expect(nbt).toEqual({
      kind: NbtType.Compound,
      name: "hello world",
      payload: {
        name: {
          kind: NbtType.String,
          name: "name",
          payload: "Bananrama",
        },
      },
    });
  });

  it("should parse the bigtest.nbt test case", () => {
    // Given
    const packet = createDummyBuffer(bigTestNbt);

    // When
    const reader = new NbtPacketReader(packet);
    const nbt = reader.readNbtCompound();

    // Then
    expect(nbt).toMatchSnapshot();
  });
});
