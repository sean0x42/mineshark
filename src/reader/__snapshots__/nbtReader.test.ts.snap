// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nbtReader should parse the bigtest.nbt test case 1`] = `
Object {
  "kind": 10,
  "name": "Level",
  "payload": Object {
    "byteArrayTest (the first 1000 values of (n*n*255+n*7)%100, starting with n=0 (0, 62, 34, 16, 8, ...))": Object {
      "kind": 7,
      "name": "byteArrayTest (the first 1000 values of (n*n*255+n*7)%100, starting with n=0 (0, 62, 34, 16, 8, ...))",
      "payload": Object {
        "data": Array [
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
          0,
          62,
          34,
          16,
          8,
          10,
          22,
          44,
          76,
          18,
          70,
          32,
          4,
          86,
          78,
          80,
          92,
          14,
          46,
          88,
          40,
          2,
          74,
          56,
          48,
          50,
          62,
          84,
          16,
          58,
          10,
          72,
          44,
          26,
          18,
          20,
          32,
          54,
          86,
          28,
          80,
          42,
          14,
          96,
          88,
          90,
          2,
          24,
          56,
          98,
          50,
          12,
          84,
          66,
          58,
          60,
          72,
          94,
          26,
          68,
          20,
          82,
          54,
          36,
          28,
          30,
          42,
          64,
          96,
          38,
          90,
          52,
          24,
          6,
          98,
          0,
          12,
          34,
          66,
          8,
          60,
          22,
          94,
          76,
          68,
          70,
          82,
          4,
          36,
          78,
          30,
          92,
          64,
          46,
          38,
          40,
          52,
          74,
          6,
          48,
        ],
        "type": "Buffer",
      },
    },
    "byteTest": Object {
      "kind": 1,
      "name": "byteTest",
      "payload": 127,
    },
    "doubleTest": Object {
      "kind": 6,
      "name": "doubleTest",
      "payload": 0.4931287132182315,
    },
    "floatTest": Object {
      "kind": 5,
      "name": "floatTest",
      "payload": 0.4982314705848694,
    },
    "intTest": Object {
      "kind": 3,
      "name": "intTest",
      "payload": 2147483647,
    },
    "listTest (compound)": Object {
      "kind": 9,
      "name": "listTest (compound)",
      "payload": Array [
        Object {
          "kind": 10,
          "name": undefined,
          "payload": Object {
            "created-on": Object {
              "kind": 4,
              "name": "created-on",
              "payload": 1264099775885n,
            },
            "name": Object {
              "kind": 8,
              "name": "name",
              "payload": "Compound tag #0",
            },
          },
        },
        Object {
          "kind": 10,
          "name": undefined,
          "payload": Object {
            "created-on": Object {
              "kind": 4,
              "name": "created-on",
              "payload": 1264099775885n,
            },
            "name": Object {
              "kind": 8,
              "name": "name",
              "payload": "Compound tag #1",
            },
          },
        },
      ],
    },
    "listTest (long)": Object {
      "kind": 9,
      "name": "listTest (long)",
      "payload": Array [
        Object {
          "kind": 4,
          "name": undefined,
          "payload": 11n,
        },
        Object {
          "kind": 4,
          "name": undefined,
          "payload": 12n,
        },
        Object {
          "kind": 4,
          "name": undefined,
          "payload": 13n,
        },
        Object {
          "kind": 4,
          "name": undefined,
          "payload": 14n,
        },
        Object {
          "kind": 4,
          "name": undefined,
          "payload": 15n,
        },
      ],
    },
    "longTest": Object {
      "kind": 4,
      "name": "longTest",
      "payload": 9223372036854775807n,
    },
    "nested compound test": Object {
      "kind": 10,
      "name": "nested compound test",
      "payload": Object {
        "egg": Object {
          "kind": 10,
          "name": "egg",
          "payload": Object {
            "name": Object {
              "kind": 8,
              "name": "name",
              "payload": "Eggbert",
            },
            "value": Object {
              "kind": 5,
              "name": "value",
              "payload": 0.5,
            },
          },
        },
        "ham": Object {
          "kind": 10,
          "name": "ham",
          "payload": Object {
            "name": Object {
              "kind": 8,
              "name": "name",
              "payload": "Hampus",
            },
            "value": Object {
              "kind": 5,
              "name": "value",
              "payload": 0.75,
            },
          },
        },
      },
    },
    "shortTest": Object {
      "kind": 2,
      "name": "shortTest",
      "payload": 32767,
    },
    "stringTest": Object {
      "kind": 8,
      "name": "stringTest",
      "payload": "HELLO WORLD THIS IS A TEST STRING ÅÄÖ!",
    },
  },
}
`;
